plugins/se/se.cpp

```diff
@@ -0,0 +1,321 @@
+	if (isPlayerDead) {
+		context.clear();
+		identity.clear();
+
+		for (uint8_t i = 0; i < 3; ++i) {
+			avatarPos[i] = avatarFront[i] = avatarTop[i] = cameraPos[i] =  cameraFront[i] = cameraTop[i] = 0.0f;
+		}
+
+		return true;
+	}
+
+	float rotation[3];
+	if (!peekProc(localPlayer + rotationOffset, rotation, sizeof(rotation))) {
+		return false;
+	}
+
+	float originPosition[3];
+	if (!peekProc(localPlayer + originPositionOffset, originPosition, sizeof(originPosition))) {
+		return false;
+	}
+
+	float eyesPositionOffset[3];
+	if (!peekProc(localPlayer + eyesPositionOffsetOffset, eyesPositionOffset, sizeof(eyesPositionOffset))) {
+		return false;
+	}
+
+	NetInfo ni;
+	if (!peekProc(localClient + netInfoOffset, ni)) {
+		return false;
+	}
+
+	std::string host;
+
+	// 0: NA_NULL
+	// 1: NA_LOOPBACK
+	// 2: NA_BROADCAST
+	// 3: NA_IP
+	switch (ni.type) {
+		case 3: {
+			std::ostringstream ss;
+			for (uint8_t i = 0; i < 4; ++i) {
+				ss << std::to_string(ni.ip[i]);
+				if (i < 3) {
+					ss << ".";
+				}
+			}
+
+			ss << ":" << networkToHost(ni.port);
+			host = ss.str();
+			break;
+		}
+		default:
+			return true;
+	}
+
+	// 40 is the size of the char array in the engine's code
+	char map[40];
+	if (!peekProc(localClient + levelNameOffset, map, sizeof(map))) {
+		map[0] = '\0';
+	}
+
+	// Begin context
+	std::ostringstream ocontext;
+	ocontext << " {\"Host\": \"" << host << "\"}"; // Set context with server's IP address and port.
+	context = ocontext.str();
+	// End context
+
+	// Begin identity
+	std::wostringstream oidentity;
+	oidentity << "{";
+
+	// Map
+	escape(map, sizeof(map));
+	if (strcmp(map, "") != 0) {
+		oidentity << std::endl << "\"Map\": \"" << map << "\""; // Set map name in identity.
+	} else {
+		oidentity << std::endl << "\"Map\": null";
+	}
+
+	oidentity << std::endl << "}";
+	identity = oidentity.str();
+	// End identity
+
+	// Mumble | Game
+	// X      | X
+	// Y      | Z
+	// Z      | Y
+	//
+	// originPosition corresponds to the origin of the player model,
+	// usually the point (i.e. feet) where it touches the ground.
+	// eyesPositionOffset is the difference between the eyes and the origin position.
+	// We use the eyes' position because they are closer to the mouth.
+	avatarPos[0] = originPosition[0] + eyesPositionOffset[0];
+	avatarPos[1] = originPosition[2] + eyesPositionOffset[2];
+	avatarPos[2] = originPosition[1] + eyesPositionOffset[1];
+
+	anglesToVectors(rotation, avatarFront, avatarTop);
+
+	// Sync camera vectors with avatar ones.
+	//
+	// This is not ideal, but it's not an issue unless 3rd person mode is enabled.
+	// The command to set it is "thirdperson" and requires "sv_cheats" to be enabled.
+	// For reference, the functions that return what we need are MainViewOrigin() and MainViewAngles().
+	for (uint8_t i = 0; i < 3; ++i) {
+		cameraPos[i] = avatarPos[i] /= 39.37f; // Convert from inches to meters.
+		cameraFront[i] = avatarFront[i];
+		cameraTop[i] = avatarTop[i];
```

Comment:
Why is only one of these values converted to meters? I would have assumed that if one entry in `avatarPos` is in inches, then the others should also be in inches, shouldn't they?

Reply:
`cameraPos` and `avatarPos` store the position, 3 coordinates that represent it in inches.
`cameraFront` and `avatarFront` store the front vector, 3 coordinates that can have a value from -1 to 1.
`cameraTop` and `avatarTop` store the top vector, 3 coordinates that can have a value from -1 to 1.

Reply:
Ah yeah. I misread the code before. :facepalm: